client/src/js/services/Snapshot.js

```diff
@@ -0,0 +1,175 @@
+
+angular.module('bhima.services')
+.service('SnapshotService', SnapshotService);
+
+SnapshotService.$inject = ['$uibModal'];
+
+function SnapshotService($uibModal) {
+
+  var service = this;
+  service.dataUriToFile = dataUriToFile;
+  service.openWebcamModal = openWebcamModal;
+
+  function openWebcamModal() { 
+    return $uibModal.open({
+      templateUrl: 'modules/templates/bhSnapShot.html',
+      controller: snapshotController,
+      backdrop     : 'static',
+      animation    : false,
+      size : 'lg',
+    }).result;
+  }
+
+  // convert the data_url to a file object
+  function dataUriToFile(dataUri, fileName, mimeType) {
+    return (fetch(dataUri)
+      .then(function(res){return res.arrayBuffer();})
+      .then(function(buf){return new File([buf], fileName, {type:mimeType});})
+    );
+  }
+
+  return service;
+}
+
+// the controler for this service
+
+angular.module('bhima.controllers')
+.controller('snapshotController', snapshotController);
+
+snapshotController.$inject = ['$scope', '$uibModalInstance'];
+
+function snapshotController($scope, $uibModalInstance) {
+  'use strict';
+  var _video = null, patData = null;
+
+  $scope.showDemos = false;
+  $scope.edgeDetection = false;
+  $scope.mono = false;
+  $scope.invert = false;
+
+  $scope.hasDataUrl = false;
+  // help to display the saving botton
+
+  $scope.patOpts = { x: 0, y: 0, w: 25, h: 25};
+
+  // Setup a channel to receive a video property
+  // with a reference to the video element
+  // See the HTML binding in main.html
+  $scope.channel = {};
+
+  $scope.webcamError = false;
+  $scope.onError = function (err) {
+    $scope.$apply(
+      function () {
+        $scope.webcamError = err;
+      }
+    );
+  };
+
+  $scope.onSuccess = function () {  
+  // The video element contains the captured camera data
+    _video = $scope.channel.video;
+    $scope.$apply(function() {
+      $scope.patOpts.w = _video.width;
+      $scope.patOpts.h = _video.height;
+      $scope.showDemos = true;
+    });
+  };
+
+  /**
+   * Make a snapshot of the camera data and show it in another canvas.
+   */
+  $scope.makeSnapshot = function makeSnapshot() {  
+    if (_video) {
+      var patCanvas = document.querySelector('#snapshot');
+      if (!patCanvas) return;
+
+        patCanvas.width = _video.width;
+        patCanvas.height = _video.height;
+        var ctxPat = patCanvas.getContext('2d');
+        var idata = getVideoData($scope.patOpts.x, $scope.patOpts.y, $scope.patOpts.w, $scope.patOpts.h);
+        ctxPat.putImageData(idata, 0, 0);
+        sendSnapshotToServer(patCanvas.toDataURL());
+        patData = idata;
+      }
+  };
+
+  /**
+   * Redirect the browser to the URL given.
+   * Used to download the image by passing a dataURL string
+   */
+  $scope.downloadSnapshot = function downloadSnapshot(dataURL) {
+    window.location.href = dataURL;
```

Comment:
This function will be easier to test if you use angular's `$window` instead of the native `window`.